browser/brave_vpn/dns/brave_vpn_dns_observer_service.cc

```diff
@@ -0,0 +1,277 @@
+                        base::SPLIT_WANT_NONEMPTY);
+  for (const auto& it : urls) {
+    if (GURL(it).is_valid())
+      return true;
+  }
+  return false;
+}
+
+std::string GetFilteredProvidersForCountry() {
+  namespace secure_dns = chrome_browser_net::secure_dns;
+  // Use default hardcoded servers for current country.
+  auto providers = secure_dns::ProvidersForCountry(
+      secure_dns::SelectEnabledProviders(net::DohProviderEntry::GetList()),
+      country_codes::GetCurrentCountryID());
+  for (const net::DohProviderEntry* entry : net::DohProviderEntry::GetList()) {
+    if (entry->provider != kBraveVpnDnsProvider)
+      continue;
+    net::DnsOverHttpsConfig doh_config({entry->doh_server_config});
+    return doh_config.ToString();
+  }
+  NOTREACHED() << "Should not be reached as we expect " << kBraveVpnDnsProvider
+               << " is available in the default list.";
+  return std::string("1.1.1.1");
+}
+
+std::string GetDoHServers(const std::string* user_servers) {
+  return user_servers && !user_servers->empty()
+             ? *user_servers
+             : GetFilteredProvidersForCountry();
+}
+
+gfx::NativeWindow GetAnchorBrowserWindow() {
+  auto* browser = chrome::FindLastActive();
+  return browser ? browser->window()->GetNativeWindow()
+                 : gfx::kNullNativeWindow;
+}
+
+}  // namespace
+
+BraveVpnDnsObserverService::BraveVpnDnsObserverService(
+    PrefService* local_state,
+    PrefService* profile_prefs,
+    DnsPolicyReaderCallback callback)
+    : policy_reader_(std::move(callback)),
+      local_state_(local_state),
+      profile_prefs_(profile_prefs) {
+  DCHECK(profile_prefs_);
+  DCHECK(local_state_);
+
+  net::NetworkChangeNotifier::AddDNSObserver(this);
+  pref_change_registrar_.Init(local_state);
+  pref_change_registrar_.Add(
+      ::prefs::kDnsOverHttpsMode,
+      base::BindRepeating(&BraveVpnDnsObserverService::OnDNSPrefChanged,
+                          weak_ptr_factory_.GetWeakPtr()));
+  pref_change_registrar_.Add(
+      ::prefs::kDnsOverHttpsTemplates,
+      base::BindRepeating(&BraveVpnDnsObserverService::OnDNSPrefChanged,
+                          weak_ptr_factory_.GetWeakPtr()));
+}
+
+BraveVpnDnsObserverService::~BraveVpnDnsObserverService() {
+  net::NetworkChangeNotifier::RemoveDNSObserver(this);
+}
+
+void BraveVpnDnsObserverService::OnDNSChanged() {
+  // Re-create service for reading config once becasue the service keeps
+  // the callback and cannot read multiple times.
+  dns_config_service_ = net::DnsConfigService::CreateSystemService();
+  dns_config_service_->ReadConfig(
+      base::BindRepeating(&BraveVpnDnsObserverService::OnSystemDNSConfigChanged,
+                          weak_ptr_factory_.GetWeakPtr()));
+}
+
+void BraveVpnDnsObserverService::OnSystemDNSConfigChanged(
+    const net::DnsConfig& config) {
+  system_dns_config_ = config;
+  dns_config_service_.reset();
+}
+
+bool BraveVpnDnsObserverService::IsDNSSecure(
```

Comment:
this is not right, we should be checking `net::ReadWinSystemDnsSettings` to see if the current dns provider is already using doh